### PR TITLE
Brain large benchmark

### PR DESCRIPTION
**Pull request by Edouard360**
Brain Large preprocessing was until now, a bit different from Romain's. (In particular, batch indices weren't taken into account). 

Romain's exact preprocessing is at c13928e.If we care exactly about having same train/test dataset, we can look at this previous commit. 

Otherwise, the same results are also reproducible taking the subsample size: 60000 in Romain's notebook.

```
gene_dataset = BrainLargeDataset(subsample_size=60000)
run_benchmarks(gene_dataset, n_epochs=250, use_batches=False, tt_split=0.8)
```

This will give train/test 48000/12000 ~= 50000/10000 in Romain's notebook. The same final likelihood score + imputation score are achieved  on test set : **LL = 1148, MAE = 1.07**

It takes 1119 sec for the 250 epochs. I think Romain does only 120 epochs, and indeed, after 40 epochs we reach the final log likelihood so we could at least reduce that training time to ~180 sec = 3min. with early stopping.

As for the Cortex Dataset, 250 epochs of training give  **LL = 1197, MAE = 2.18**, which is good for MAE, and a little above for LL test (~ 1180 in Romain's preprint), even using **adapt_encoder** new function. But imo with a little tuning (learning rate, batch size), Cortex result's are easily reproducible.

In terms of timing / GPU usage, **nvidia-smi => 40/50%** due to the `data_loader.next` procedure.

### CODE REVIEW

File: run_benchmarks.py [hunk 1]
```diff
@@ -14,12 +13,9 @@
     parser.add_argument("--nobatches", action='store_true', help="whether to ignore batches")
     parser.add_argument("--nocuda", action='store_true',
                         help="whether to use cuda (will apply only if cuda is available")
+    parser.add_argument("--benchmark", action='store_true',
+                        help="whether to use cuda (will apply only if cuda is available")
     args = parser.parse_args()
     gene_dataset = load_datasets(args.dataset)
```
Comment (jeff-regier) [inline run_benchmarks.py]: I like that these changes move more of the logic to `run_benchmarks`. Can we move the call to `load_dataset` to `run_benchmarks` too?

File: scvi/dataset/const.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
+string_10x = """E18 20161004 Neurons Sample 64 Fraction of Reads Kept	87.1%
```
Comment (jeff-regier) [inline scvi/dataset/const.py]: Grouping variables into files by data type isn't such a clear organizing principle, e.g., all the ints in one file, all the consts in another file.
If I understand what you're doing here, then maybe better to put this where it's used: in `brain_large.py`. Or do you have in mind using this for `brain_small` too? Then maybe put all of them--- `BrainLargeDataset`, `BrainSmallDataset`, and `string_10x`---in a file called `ten_x.py`

Comment (jeff-regier) [reply scvi/dataset/const.py]: Also, better to parse them again of time, and keep all the irrelevant stuff out of the repo. e.g., `neurons_sample = [63, 62, ...]`

File: scvi/utils.py [hunk 3]
```diff
@@ -23,5 +25,21 @@ def wrapper(*args, **kwargs):
         return wrapper
 
 
+class pickle_result:
```
Comment (jeff-regier) [inline scvi/utils.py]: What are we saving with pickle? (If it's neural net parameters, better use `torch.save` and `torch.load`.) e.g.

```
    vae.load_state_dict(torch.load("params.dat"), strict=False)
    train_module(vae, ds, epochs=args.epochs)
    torch.save(vae.state_dict(), "params.dat")
```

Comment (Edouard360) [reply scvi/utils.py]: I used that as a convenient debugging tools when benchmarking on brain_large with different hyperparameters. 

It can take ~5min to preprocess data in instance, so by adding:

```
@pickle_result('data/tmp')
def preprocess(self):
```

I am not pushing this line, but this is saving time. Does this makes sense ? (I can remove and only keep locally otherwise)

Comment (jeff-regier) [reply scvi/utils.py]: I see. It's better to keep debugging code locally in a file but it's not a big deal for now since only a few of us our reading the code. Once we move from scvi-dev to scVI, we should keep all the debugging code out of the repo, to make it easier for people to read the code.

**Comment by jeff-regier** [comment]
@Edouard360 Can you increase the GPU utilization to ~100% by setting `num_workers=4`, and adding
```
torch.multiprocessing.set_start_method("spawn")
```
so we don't get a seg fault?

**Comment by jeff-regier** [comment]
Does this run on all of brain_large?

**Comment by Edouard360** [comment]
720 genes and 50000/10000 cells for train/test

doesn't these settings slow down performance ? I have tried torch.multiprocessing.set_start_method("spawn") combined with num_workers=4, but, at least for Cortex and the brain large subsample, it does worse in computation time. I will try again

File: scvi/utils.py (hunk 3 above)
Comment (Edouard360) [reply scvi/utils.py]: ok !

**Comment by jeff-regier** [comment]
If it turns out that indexing into arrays other than `self.X` contributes significantly to runtime, we might explore packing everything for a cell into a "structured array":

https://docs.scipy.org/doc/numpy-1.14.0/user/basics.rec.html

That way all the metadata for each cell is stored contiguously in memory, greatly reducing the look up time.

**Comment by Edouard360** [comment]
I will investigate the performance gain from that approach ! Thank you for the idea !

**Comment by jeff-regier** [comment]
Great, thanks for looking into it. One other question to explore, assuming we're I/O bound, is what type of bandwidth is limiting: main memory to/from CPU, or CPU to GPU. The script I suggested above tests just the former---I think that's most likely the problem. If it's not, i.e., you can iterate through the whole dataset (without transferring anything to the GPU, and without doing any computation) much faster than one epoch, then it's transfer to the GPU that's limiting. We could reduce transfer to the GPU by transferring sparse Pytorch tensors: 

https://pytorch.org/docs/master/sparse.html

Basically the idea is to call `.cuda()` on a sparse tensor to get it to the GPU. Then, on the GPU, call `to_dense()`.

**Comment by Edouard360** [comment]
```
Original setup :  6.614995070998702
SyntheticNamedDataset1  :  2.031112746000872
SyntheticNamedDataset2  :  2.3277608979988145
SyntheticConcatenatedDataset1  :  2.307500159000483
SyntheticConcatenatedDataset2  :  0.7280285240012745
SyntheticConcatenatedDataset3  :  1.350101240001095
```

Original setup is the way we currently iterate.

These results come from running the following gist:

https://gist.github.com/Edouard360/4615b4dc35b58c0a2a89be57e7b165d5

- collate_fn function helps significantly. 
- Also, faster to keep indexing in numpy as long as possible. 
- Finally, though a structured array reduce the time compared to what we had before, we can even further reduce it if we keep one same matrix.

**Comment by Edouard360** [comment]
[UPDATE]

```
Original setup :  6.868990107002901
CUDA ASYNC: Original setup :  10.628782384999795
CUDA: Original setup :  7.163563253998291

SyntheticNamedDataset1  :  2.1348808510010713
CUDA ASYNC: SyntheticNamedDataset1  :  6.104878799000289
CUDA : SyntheticNamedDataset1  :  2.581230169002083

SyntheticNamedDataset2  :  2.3647528359942953
CUDA ASYNC: SyntheticNamedDataset2  :  2.6241530170009355
CUDA : SyntheticNamedDataset2  :  2.8522622250020504

SyntheticConcatenatedDataset1  :  2.3768859440024244
CUDA ASYNC: SyntheticConcatenatedDataset1  :  2.621973554996657
CUDA : SyntheticConcatenatedDataset1  :  2.8486932910018368

SyntheticConcatenatedDataset2  :  0.7307170310014044
CUDA ASYNC: SyntheticConcatenatedDataset2  :  0.9630912069987971
CUDA : SyntheticConcatenatedDataset2  :  1.182853446000081

SyntheticConcatenatedDataset3  :  1.3636612720001722
# cuda not possible (arrays not contiguous)
```

**Comment by jeff-regier** [comment]
Really interesting results, thanks for getting to the bottom of the problem. It's pretty ugly storing metadata in the data matrix as extra columns---I was really hoping the structured array approach would work out. But hard to argue with the numbers! Given how inefficiently Python does some things, I suppose we should go with the approach in `SyntheticConcatenatedDataset2`---probably speed is more important than clarity here since I/O is a bottleneck.

**Comment by Edouard360** [comment]
I agree this is unfortunate, and it would have been cleaner with structured arrays:

One last experiment I want to mention is on Synthetic Dataset with parameters `batch_size=5000, nb_genes=1000`

```
# Current Implementation
Train loop : 30.04613575300027
Loading async : 9.131983047991525
Loading : 9.719744369009277

# Current Implementation, but everything to numpy - same dataloader pipeline [get_item + default collate_fn])
Train loop : 31.160587372003647
Loading async : 8.933563602004142
Loading : 9.576901852000447

# SyntheticNamedDataset2
batch_size=5000, nb_genes=1000
Train loop : 23.48242518200277
Loading async : 3.2243813649984077
Loading : 3.62294194700371

# SyntheticNamedDataset2
batch_size=5000, nb_genes=100
Train loop : 22.154694396005652
Loading async : 2.765005180001026
Loading : 3.007767760005663

# SyntheticConcatenatedDataset2
batch_size=5000, nb_genes=1000
Train loop : 22.680343132000417
Loading async : 1.196391092002159
Loading : 1.388669376996404

# SyntheticConcatenatedDataset2
batch_size=5000, nb_genes=100
Train loop : 21.649731174999033
Loading async : 1.0033584699995117
Loading : 1.2030913829948986
```

I will adapt code and commit so we use SyntheticConcatenatedDataset2, but it looks though like the overhead of SyntheticNamedDataset2 is, at least, much less significant than what we had before. It could be further reduced (maybe ?) with a more clever collate_fn - **that parralels the concatenation of the elements**, because I suspect that the following chunk of code is suboptimal...

```
    def collate_fn(batch):
        counts, local_mean, local_var, batch_indices, labels = [], [], [], [], []
        for b in batch:
            counts += [b[0]]
            local_mean += [b[1]]
            local_var += [b[2]]
            batch_indices += [b[3]]
            labels += [b[4]]

        return [torch.FloatTensor(np.concatenate(l))
                for l in [counts, local_mean, local_var, batch_indices, labels]]
```

**Comment by jeff-regier** [comment]
Yes, "for" loops in Python are pretty inefficient (compared to C or Cython or Julia). With the "everything in one matrix" approach, the loops are done by C code.

Also, it's generally faster to preallocate arrays, rather than to grow them dynamically. That's easily fixed,
e.g., ```
counts = np.zeros(len(batch))

for i in len(batch):
    counts[i] = batch[i][0]
```
But I'm not sure about how to get rid of the for loops, which may be more significant.

**Comment by jeff-regier** [comment]
One other idea: What about defining, in the Dataset class,

```
def __getitem__(idx):
    return idx
```

And then doing all the fetching in `collate_fn` ? e.g.,

```
def collate_fn(batch):
   indexes = np.array(batch)
   return self.counts[indexes], self.local_mean[indexes], ...
```

Might be a lot faster because we'd be fetching all the rows at once.

**Comment by Edouard360** [comment]
Haha, that sounds like a nice trick ! I try it out
